crates/hir_ty/src/diagnostics/expr.rs

```diff
@@ -78,6 +79,12 @@ impl<'a, 'b> ExprValidator<'a, 'b> {
         let body_expr = &body[body.body_expr];
         if let Expr::Block { tail: Some(t), .. } = body_expr {
             self.validate_results_in_tail_expr(body.body_expr, *t, db);
+        } else {
+            if let Expr::Block { statements, .. } = body_expr {
+                if let Some(Statement::Expr(id)) = statements.last() {
+                    self.validate_missing_tail_expr(body.body_expr, *id, db);
+                }
+            }
         }
```

Comment:
```suggestion
        if let Expr::Block { statements, tail, .. } = body_expr {
            if let Some(t) = tail {
                self.validate_results_in_tail_expr(body.body_expr, *t, db);
            } else if let Some(Statement::Expr(id)) = statements.last() {
                self.validate_missing_tail_expr(body.body_expr, *id, db);
            }
        }
```

Looks a bit nicer IMO.